File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.go
func findCheapestPrice(n int, flights [][]int, src int, dst int, k int) int {
    price := make([]int, n)
    for i := 0; i < n; i++ {
        price[i] = -1
    }

    price[src] = 0

    for stop := 0; stop <= k; stop++ {
        temp := make([]int, n)
        copy(temp, price)

        for _, flight := range flights {
            from := flight[0]
            to := flight[1]
            p := flight[2]

            if price[from] == -1 {
                continue
            }

            total := price[from] + p

            if temp[to] == -1 || temp[to] > total {
                temp[to] = total
            }
        }

        price = temp
    }

    return price[dst]
}

File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.rs
impl Solution {
    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        // result := make([]int, n)
        let mut result: Vec<i32> = vec![-1; n as usize];
        // println!("result: {:?}", result);

        result[src as usize] = 0;
        // println!("result: {:?}", result);

        for i in 0..k + 1 {
            let mut temp: Vec<i32> = result.clone();

            for flight in flights.iter() {
                let from: i32 = flight[0];
                let to: i32 = flight[1];
                let price: i32 = flight[2];

                if result[from as usize] == -1 {
                    continue;
                }

                if temp[to as usize] == -1 || temp[to as usize] > result[from as usize] + price {
                    temp[to as usize] = result[from as usize] + price;
                }

                // println!("i: {}, temp: {:?}", i, temp);
            }

            result = temp;
        }

        return result[dst as usize];
    }
}